### PR TITLE
Fix issue 19278 - extern(C++, "name") doesn't accept expressions

### PR DESCRIPTION
**Pull request by TurkeyMan**
Accept an expression, attempt to resolve in semantic.
Ambiguity with scoped namespace identifier can be defeated with parens.

### CODE REVIEW

File: src/dmd/dsymbolsem.d
Comment (TurkeyMan) [inline src/dmd/dsymbolsem.d]: Can I get guidance on this block? I'm not sure where/how it should be placed inside this function... and is there anything that should be done after the late-assignment of the symbol name that might have been missed prior?

**Comment by WalterBright** [comment]
Does this mean https://github.com/dlang/dmd/pull/8781 is abandoned? Should it be Phantom Zoned, or just closed?

**Comment by TurkeyMan** [comment]
No, I think that is complementary; being able to compose strings this way is a lot less useful if you can't compose a string with `::` in it.

I'll merge them if it's accepted, but I wanted to review the detail in this one as a self-contained patch, since it's more complex. I didn't want to muddy the waters.
If this approach is accepted, I think I'll also route the string-literal path through this same code to unify the surface-area of the feature, and move the `::` handling into semantic.

**Comment by TurkeyMan** [comment]
I think strings with "::" in them feels more uniform with `extern(C++, ns.ns2.ns3)` which is accepted via the scoped path. `extern(C++, "ns::ns2::ns3")` feels more like the proper complelemt, and it also lends to simpler composition expressions without involving tuples and `.expand` (which are in phobos, not druntime!).

**Comment by TurkeyMan** [comment]
If you can comment on the work-block; is it in the right location? Ordered correctly relative to the other work? Does it need any further semantic work?

**Comment by jacob-carlborg** [comment]
> I think strings with "::" in them feels more uniform with extern(C++, ns.ns2.ns3) which is accepted via the scoped path. extern(C++, "ns::ns2::ns3") feels more like the proper complelemt, and it also lends to simpler composition expressions without involving tuples and .expand (which are in phobos, not druntime!).

The reason I suggested several arguments/tuple was to avoid the need to decide between `”foo::bar”` and `”foo.bar”`. It also not necessary for the compiler to parse the string, which in my opinion feels a bit ugly. 

But I didn’t think of a need to compose the string literal.

File: src/dmd/parse.d
Comment (thewilsonator) [inline src/dmd/parse.d]: My parser knowledge is a bit limited, but shouldn't this be an assignExp?

Comment (TurkeyMan) [reply src/dmd/parse.d]: How would an assign expression be valid here? I think `?:` is the highest level grammar applicable in this context...

**Comment by TurkeyMan** [comment]
> But I didn’t think of a need to compose the string literal.

Fair enough. That was one of primary motives for making it a string ;)

So, I'm feeling '::' because that's a cut & paste of the C++ namespace that it's linking against, and it also leaves no room for confusion that it might be a D identifier.

**Comment by WalterBright** [comment]
> I know it's not technically difficult, it just feels wrong.

I agree, it's hackish to have a "language within a language".

**Comment by TurkeyMan** [comment]
Ummm. C++ support is *definitely* a language within a language to some quite ambitious extent... this expression describes a C++ namespace for mangling purposes, and that should be a string such that it may be any valid C++ namespace. The only grammatical way to express a C++ namespace in D is within a string, and as a bonus by doing so, we're not leaking any possibility for misconception that the thing is anything other than a C++ namespace to the user... I think it's perfect.
If it used '.' instead of '::', or not-a-string, it starts to look like D syntax, and it's definitely not D syntax, and shouldn't be confuse-able as such.
I really think this design has a massive clarity advantage. Users can't misunderstand what they're looking at.

**Comment by wilzbach** [comment]
Maybe I didn't catch the full discussion, but why is the parenthesis required?

**Comment by TurkeyMan** [comment]
It's not _required_... any expression that doesn't *start* with an identifier is unambiguous... it's just that i figured requiring parens was a little more explicit and less susceptible to surprise.

**Comment by wilzbach** [comment]
It just looks/looked like an arbitrary limitation to me and if I would look at code that uses it I would probably wonder why the author used the unnecessary parenthesis.
It just reminded me about a silliness in JavaScript's destructing rules where the parentheses are required in the last line:

```js
{a: b} = {a: 42};
// b is 42
let obj = {};
({a: obj.b} = {a: 12}); 
// obj now has a member b with the value 12
```

Anyhow, AFAICT that's not the main concern here.

**Comment by TurkeyMan** [comment]
There's a potential ambiguity between the names scope and a string expression.
I went with parens precisely because of precedent of the kind that you just described ;)
C++ has those same use-parents-to-break-ambiguity tricks all over the place.
The requirement for parens can be relaxed if named scopes are lifted into a distinct feature in the future (which I think would be ideal; named scopes should be a feature that stands in its own right if it should exist).

**Comment by TurkeyMan** [comment]
`test/stdcpp/Makefile` is in the patch. That triggers the build. You can see the `.cpp` compile in the build log.

**Comment by jacob-carlborg** [comment]
> If it used '.' instead of '::', or not-a-string, it starts to look like D syntax, and it's definitely not D syntax, and shouldn't be confuse-able as such.

It *is* D syntax. I'm mean, it is **D** after all.

File: src/dmd/dsymbolsem.d
Comment (TurkeyMan) [inline src/dmd/dsymbolsem.d]: I'm not sure what to write here to fix up the parent/scope pointers...

**Comment by TurkeyMan** [comment]
No... it's a string, that contains a C++ identifier to be used for C++ mangling. The D syntax is that it accepts a string of a valid C++ identifier for mangling. It's not valid D identifiers, and it shouldn't be possible to confuse it as such.

File: test/fail_compilation/cppmangle.d
Comment (thewilsonator) [inline test/fail_compilation/cppmangle.d]: You need to update the line number and message, from the autotester:
```
expected:
----
fail_compilation/cppmangle.d(10): Error: string expected following `,` for C++ namespace, not `)`
----
actual:
----
fail_compilation/cppmangle.d(8): Error: found `,` when expecting `)`
fail_compilation/cppmangle.d(8): Error: declaration expected, not `)`
----
```
so this would become
```
fail_compilation/cppmangle.d(9): Error: found `,` when expecting `)`
fail_compilation/cppmangle.d(9): Error: declaration expected, not `)`
```

File: test/fail_compilation/cppmangle2.d
Comment (thewilsonator) [inline test/fail_compilation/cppmangle2.d]: The line numbers here area incorrect.

File: src/dmd/dsymbolsem.d
Comment (thewilsonator) [reply src/dmd/dsymbolsem.d]: Please spilt that loop into the else case followed by a loop over the rest of the namespaces.

Comment (thewilsonator) [reply src/dmd/dsymbolsem.d]: So, the AST at the start looks like 
```
NSpace foo::bar::baz
+- Dsymbols members
```
After the end of the above it should look like (modulo `s.addMember(sc, sc.scopesym)`)
```
NSpace foo
+-Dsymbols members
+- NSpace bar
 +- NSpace baz
```
and you want it to end up looking like 
```
NSpace foo
+- NSpace bar
 +- NSpace baz
  +- Dsymbols members
```
In which case instead of that I'd validate all the namespace identifiers _first_, and then pull the ol' switcharoo:
```
Enclosing Scope (e.g. module)
+-NSpace foo::bar::baz
 +- Dsymbols members
```
split,  do _not_ insert into enclosing scope ->
```
NSpace bar::baz (parentless)
Enclosing Scope
+- NSpace foo
 +- Dsymbols members
```
expand ->
```
NSpace bar (parentless)
+- NSpace baz
Enclosing Scope
+- NSpace foo
 +- Dsymbols members
```
rotate reassign `ident`s ->
```
NSpace foo (parentless)
+- NSpace bar
Enclosing Scope
+- NSpace baz 
 +- Dsymbols members
```
pull the switcharoo (reparenting insert) ->
```
Enclosing Scope
+- NSpace foo
 +- NSpace bar
  +-NSpace baz 
   +- Dsymbols members
```

This means that the original namespace keeps all the members and the only reference to the original (because its a tree), from the enclosing scope, is overwritten. then you can `addMember` the intervening namespaces.

Comment (TurkeyMan) [reply src/dmd/dsymbolsem.d]: Right... see, logically it makes sense... but i have no idea what the code that implements that might look like.
Those functions `addMember`, `setScope`, `importAll`, and what they mean relative to the 'current' scope (which is a concept that still isn't clear), and when they should all be called... and why. I can't imagine how to implement that sequence of operations with the primitives at hand.

File: test/compilable/cppmangle.d
Comment (jacob-carlborg) [inline test/compilable/cppmangle.d]: Did you implement support for `extern (C++, (foo))`?

Comment (TurkeyMan) [reply test/compilable/cppmangle.d]: yeah i'm trying to work that out... i think the moment in semantic where i inserted my code is too early to perform symbol lookups :/
what's the proper earliest moment in compilation where ctfe is able to resolve symbols? I think I need to move this code to there...

File: src/dmd/dsymbolsem.d
Comment (thewilsonator) [reply src/dmd/dsymbolsem.d]: Not sure, try asking on slack.

Comment (TurkeyMan) [reply src/dmd/dsymbolsem.d]: Please feel free to submit a patch if you think you know how to implement this logic ;)

**Comment by WalterBright** [comment]
"C++ also has those same use-parens-to-break-ambiguity tricks all over the place, among others. It's pretty common"

Yes, I know C++ does that, but it has a terrible smell about it.

**Comment by WalterBright** [comment]
I also do not like the :: syntax. It should instead be a tuple of strings.

**Comment by TurkeyMan** [comment]
> Yes, I know C++ does that, but it has a terrible smell about it.

I agree, but it's ideally temporary, until named scopes are moved into their own syntax, and the 2 features can be used in conjunction.

> I also do not like the :: syntax. It should instead be a tuple of strings.

I kinda like the '::', because it's clear that it's a C++ namespace, and you can cut&paste the namespace from C++... but  don't really mind either way.
I do feel like a tuple is harder to compose, because interacting with tuples is more advanced language mischief than strings. It's also harder to implement.

But regardless, I think I've reached about the extent of my ability with this PR without investing extensive additional time which I don't have. I've already put more time than I have available into this stuff, and it's coming at the expense of enough sleep, which is not sustainable.

This is working, tests are passing... if you want to change how it's expressed, please by my guest.

**Comment by jacob-carlborg** [comment]
> It should instead be a tuple of strings.

That's how the current implementation works. But I guess those can only be string literals.

**Comment by thewilsonator** [comment]
Indeed this is intended to solve the problem of c++ 11 namespaces (`::std` vs `::std::__1`), which requires an expression to correctly decide which one to use.

**Comment by TurkeyMan** [comment]
It is tuples now... is that all?

**Comment by TurkeyMan** [comment]
...and the tests pass!

**Comment by TurkeyMan** [comment]
This is blocking the STL stuff which has been sitting formant for 2 months...

**Comment by Geod24** [comment]
Well it is not really blocking it. You could do without, it's just ugly.
IIUC, one way to do it without too much duplication would be:

```dlang
private enum Code = q{
  //...
};
version (X)
   extern(C++, `std`) { extern(C++, `__1`) { mixin(Code); } }`
else
   extern(C++, `std`) { mixin(Code); }`
```

Correct ?

**Comment by TurkeyMan** [comment]
No. Not even ever.

**Comment by Geod24** [comment]
Care to elaborate ?

**Comment by TurkeyMan** [comment]
I'm not in the business of writing code that makes C++ look great by contrast. I won't merge a gigantic module into druntime wrapped in a string.

**Comment by TurkeyMan** [comment]
And this is done now. I spent weeks on it.

**Comment by Geod24** [comment]
> I'm not in the business of writing code that makes C++ look great by contrast. I won't merge a gigantic module into druntime wrapped in a string.

How does it makes C++ look great by contrast ? Yes it's wrapped in a token string, it's unfortunate but far from being as horrible as you try to make it sound.
As long as this is limited to Druntime and a very few minor occurrences in user code, it's IMO a completely acceptable temporary solution.

Regarding this PR: As @WalterBright pointed out, having parenthesis disambiguate between two syntaxes is pretty ugly. And that's not how we'll make the code look better than C++ for sure.
In addition, this change is a convenience feature: there's nothing that this PR enables which wasn't possible before, albeit in an uglier way.

I think this could ultimately be the direction we want to go, as accepting expressions is both consistent with `pragma(mangle)` and better feature wise than what `extern(C++, identifier)` does, but unfortunately the later is already in the language and released, and that's a situation we need to consider before piling up more features and syntaxes on top of each other.

**Comment by TurkeyMan** [comment]
You lose syntax highlighting, you lose autocomplete, you lose go-to definition, you lose ability to debug... I value all those things extremely highly.

**Comment by TurkeyMan** [comment]
Walter accepted the previous PR before it was finished baking. I'm finishing it up.
Parens aren't required, they're only necessary in the case you need to disambiguate (ie, very rarely). As raw strings, this is exactly what was there's before.

**Comment by jacob-carlborg** [comment]
Can we deprecate `extern(C++, Namespace)` and only support `extern(C++, "Namespace")`? Then there would be a conflict. A workaround can be used until the feature is removed.

**Comment by Geod24** [comment]
> You lose syntax highlighting

`q{}` blocks are syntax highlighted. [Even Github recognizes it](https://github.com/vibe-d/vibe.d/blob/6295e3a63508c9ef3445a5039a756e0f2d973873/web/vibe/web/rest.d#L1617-L1641).

> you lose ability to debug

Since they are just `extern` definitions, the debugger doesn't care.

> you lose autocomplete, you lose go-to definition

True for client code, but truly manageable for the most common std types.

I'm not attempting to block this PR, but I would prefer us to gather experience before we put in changes to the syntax that don't enable any new usage.

**Comment by jacob-carlborg** [comment]
> q{} blocks are syntax highlighted. Even Github recognizes it.

Technically GitHub does **not** recognize `q{}` blocks. It just happens to work because it doesn't recognize the `q{}` and just continues to do the syntax highlighting inside `q{}`.

**Comment by TurkeyMan** [comment]
> > You lose syntax highlighting
> 
> `q{}` blocks are syntax highlighted. [Even Github recognizes it](https://github.com/vibe-d/vibe.d/blob/6295e3a63508c9ef3445a5039a756e0f2d973873/web/vibe/web/rest.d#L1617-L1641).

Only one of my editors does it right.

> > you lose ability to debug
> 
> Since they are just `extern` definitions, the debugger doesn't care.

There's almost no `extern` in the STL bindings; that was a failed experiment.

> > you lose autocomplete, you lose go-to definition
> 
> True for client code, but truly manageable for the most common std types.

Client code is what matters!

And I'd argue that STL is one of the worst offenders; nobody knows what members they have. In C++, I would say the STL is my single greatest use for auto-complete assistance. Nobody knows how to work that thing, it's all weirdly named functions and iterators and rubbish.

> I'm not attempting to block this PR, but I would prefer us to gather experience before we put in changes to the syntax that don't enable any new usage.

This enables one of the key cases that it was intended for, and the only case I'm aware of that's in development for druntime. I'm just finishing the patch that was already merged, I see this is a completion of that task. I never imagined for a second that string expressions wouldn't be `Expressions`s.

There's definitely a future where named scopes are granted their own syntax, and the disambiguation requirement ceases to exist. In 99% of cases, just string literals are fine, and no disambiguation is required. It's really just the STL, and one other private project that I've encountered where an expression is necessary. It's not pervasive.

**Comment by 12345swordy** [comment]
> 
> 
> Can we deprecate `extern(C++, Namespace)` and only support `extern(C++, "Namespace")`? Then there would be a conflict. A workaround can be used until the feature is removed.

We going to need a DIP for that to happen. I can write up a quick draft on it.

**Comment by TurkeyMan** [comment]
> > Can we deprecate `extern(C++, Namespace)` and only support `extern(C++, "Namespace")`? Then there would be a conflict. A workaround can be used until the feature is removed.
> 
> We going to need a DIP for that to happen. I can write up a quick draft on it.

Sure. But I don't think it's important right now. I doubt anybody in the world is waiting on that.
But I'd really like to be un-blocked here...

File: src/dmd/dsymbolsem.d
Comment (rainers) [inline src/dmd/dsymbolsem.d]: Is it really necessary to call expressionSemantic again here?

File: src/dmd/dsymbolsem.d
Comment (rainers) [inline src/dmd/dsymbolsem.d]: D allows more characters in identifiers as C++, but that's probably ok.

File: src/dmd/dsymbolsem.d
Comment (rainers) [inline src/dmd/dsymbolsem.d]: `parentns.loc` and `parentns.mangleOnly` are constant here, maybe that's more obvious by using `ns` instead of `parentns`.

**Comment by 12345swordy** [comment]
Ping @WalterBright 
Is this good to go now?

File: src/dmd/dsymbolsem.d
Comment (rainers) [inline src/dmd/dsymbolsem.d]: Is this necessary? The scopesym should not change by inserting mangleOnly namespaces and it seems dangerous to insert symbols into the symbol table again. Maybe calling s.setScope(sc) is enough.

File: src/dmd/parse.d
Comment (rainers) [inline src/dmd/parse.d]: The brackets don't match the comment before the function.

File: test/fail_compilation/cppmangle.d
Comment (rainers) [inline test/fail_compilation/cppmangle.d]: Is this no longer an error or can the test be kept?

File: test/fail_compilation/ice17074.d
Comment (rainers) [inline test/fail_compilation/ice17074.d]: Why are these tests change or removed?

File: src/dmd/dsymbolsem.d
Comment (TurkeyMan) [reply src/dmd/dsymbolsem.d]: The tuple must be strings. I considered this message for a while, but felt that sticking with this message was less confusing. The message is: this expects strings. So explain that it must be a string or a tuple of strings felt too long and potentially more confusing that this message. YMMV?

File: src/dmd/dsymbolsem.d
Comment (TurkeyMan) [reply src/dmd/dsymbolsem.d]: How do I do that?

File: src/dmd/dsymbolsem.d
Comment (TurkeyMan) [reply src/dmd/dsymbolsem.d]: Oh, good catch! This must be a cut&paste fail.

File: src/dmd/dsymbolsem.d
Comment (TurkeyMan) [reply src/dmd/dsymbolsem.d]: Yeah I thought about that, but I felt that writing a new validation function that's used in exactly one place was overkill.
You're limited to C++ subset in practice anyway, because the C++ code is restricted bu its rules.

File: src/dmd/dsymbolsem.d
Comment (TurkeyMan) [reply src/dmd/dsymbolsem.d]: Okay.

File: src/dmd/dsymbolsem.d
Comment (TurkeyMan) [reply src/dmd/dsymbolsem.d]: I didn't understand this code as well as I'd like. I can try your suggestion?
I'd love an expert to fix these couple of lines, they made me a touch nervous, but the unit tests all passed, so I had moderate confidence I worked it out right...

File: src/dmd/dsymbolsem.d
Comment (TurkeyMan) [reply src/dmd/dsymbolsem.d]: Like, the precision is interesting to a compiler author, but a user just needs to know that they need to supply strings.

Comment (rainers) [reply src/dmd/dsymbolsem.d]: So maybe use the plural "string expressions"?

File: src/dmd/dsymbolsem.d
Comment (rainers) [reply src/dmd/dsymbolsem.d]: Yes, `error(ns.Loc, ...)` should work.

File: test/fail_compilation/cppmangle.d
Comment (rainers) [reply test/fail_compilation/cppmangle.d]: Ok, but it might be interesting to verify whether the tuple is allowed to have a trailing comma or not.

File: test/fail_compilation/ice17074.d
Comment (rainers) [reply test/fail_compilation/ice17074.d]: Ok, shouldn't have an effect on coverage.

File: src/dmd/parse.d
Comment (rainers) [reply src/dmd/parse.d]: You might also want to add the `extern(C++,struct)` to the comment.

File: src/dmd/dsymbolsem.d
Comment (jacob-carlborg) [reply src/dmd/dsymbolsem.d]: > To explain that it must be a string or a tuple of strings felt too long and potentially more confusing that this message

In theory it could be an array of strings instead of a tuple, which is not the same thing.

File: test/compilable/cppmangle.d
Comment (Geod24) [inline test/compilable/cppmangle.d]: Can I haz POSIX tests ?

File: test/fail_compilation/cppmangle.d
Comment (Geod24) [reply test/fail_compilation/cppmangle.d]: Agreed with Rainers here, can you add a trailing comma somewhere in your `cppmangle` tests ?

**Review by Geod24 — APPROVED** [review]
Could you:
- rebase
- squash
- provide a more sensible title to the resulting commit, e.g. `Fix issue 19278 - extern(C++, "name") doesn't accept expressions`
- Ping me when it's done

Tested it locally, it LGTM, and if it's the last thing needed for proper stdlib support, we might as well get it in at the beginning of the new release.

File: test/fail_compilation/cppmangle.d
Comment (TurkeyMan) [reply test/fail_compilation/cppmangle.d]: Are you sure? The fail is in the parser, not semantic... so it can't be in this file with the rest of these semantic errors. I need to add a whole new file to the tester just to test that one thing, which should be tested by whatever already tests `parseCondExp()`...
I can do it, I just don't know why, and I have to create a whole new file just for that one test.

**Comment by TurkeyMan** [comment]
Wtf is with semaphore, and why doesn't it ever work? The error is unhelpful.

**Comment by thewilsonator** [comment]
Ignore it, its unrelated.

File: test/fail_compilation/cppmangle.d
Comment (Geod24) [reply test/fail_compilation/cppmangle.d]: Hum, right, because you don't actually pass a tuple, so the point in moot. I guess when we fix this code (I see this state as temporary), we'll fix it. At the moment basically any usage of the feature needs to be surrounded by parenthesis.

**Comment by TurkeyMan** [comment]
Are we done here?

**Comment by thewilsonator** [comment]
Just squash once more and I think thats it.

File: test/fail_compilation/cppmangle.d
Comment (TurkeyMan) [reply test/fail_compilation/cppmangle.d]: There are expressions that aren't ambiguous with an identifier.
`extern(C++, "hello" ~ "world")` doesn't need parens for instance.

**Review by rainers — APPROVED** [review]

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**
